File: bindings/dotnet/src/Zlink/Contracts/Eventing/Timer.cs
// SPDX-License-Identifier: MPL-2.0

using System;

namespace Systems.Zlink;

/// <summary>
/// Defines the zlink timer contract.
/// </summary>
public interface IZlinkTimer : IDisposable, IAsyncDisposable
{
    /// <summary>
    /// Starts the timer.
    /// </summary>
    void Start(TimeSpan interval, ulong repeatCount);

    /// <summary>
    /// Stops the timer.
    /// </summary>
    void Stop();

    /// <summary>
    /// Receives the next available item.
    /// </summary>
    ulong? Recv(RecvFlags flags = RecvFlags.None);

    /// <summary>
    /// Registers a handler for fire.
    /// </summary>
    void OnFire(Action<IZlinkTimer, ulong> handler);

    /// <summary>
    /// Closes the resource.
    /// </summary>
    void Close();
}

File: bindings/java/src/main/java/systems/zlink/contracts/eventing/Timer.java
/* SPDX-License-Identifier: MPL-2.0 */

package systems.zlink.contracts.eventing;

import java.time.Duration;

public interface Timer extends AutoCloseable {

    public abstract void start(Duration interval, long repeatCount);

    public abstract void stop();

    public abstract long recv();

    public abstract void onFire(TimerHandler handler);

    @Override
    public abstract void close();
}
